data/market.py

```diff
@@ -0,0 +1,176 @@
+        """Executes an exact output swap.
+
+        Args:
+            amount (int): Amount of 'output_token' to swap (buy from the pool)
+            output_token (Token): The token to receive as the output of the swap.
+
+        Raises:
+            Exception: Raises and error if swap amount is too large.
+
+        Returns:
+            float: _description_
+        """
+        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
+            token=output_token, amount= amount, type="exact_output"
+        )
+
+    def exact_input_swap(self, input_token: Token, amount: int) -> float:
+        """Executes and exacpt input swap.
+
+        Args:
+            amount (int): Amount of 'input_token' to swap (sell to the pool)
+            input_token (Token): The token to send to the pool as the input of the swap.
+
+        Returns:
+            float: _description_
+        """
+        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
+            token=input_token, amount=amount, type="exact_input"
+        )
+
+    def exact_input_swap_slippage(self, input_token: Token, amount: int) -> float:
+        """Executes and exacpt input swap.
+
+        Args:
+            amount (int): Amount of 'input_token' to swap (sell to the pool)
+            input_token (Token): The token to send to the pool as the input of the swap.
+
+        Returns:
+            float: _description_
+        """
+        _, _, slippage = self.calculate_params(
+            token=input_token, amount=amount, type="exact_input"
+        )
+        return slippage
+
+    def exact_output_swap_slippage(self, output_token: Token, amount: int) -> float:
+        """Executes an exact output swap.
+
+        Args:
+            amount (int): Amount of 'output_token' to swap (buy from the pool)
+            output_token (Token): The token to receive as the output of the swap.
+
+        Raises:
+            Exception: Raises and error if swap amount is too large.
+
+        Returns:
+            float: _description_
+        """
+        _, _, slippage = self.calculate_params(
+            token=output_token, amount=amount, type="exact_output"
+        )
+        return slippage
+
+    def calculate_params(self, token: Token, amount: int, type: str = "exact_output") -> tuple[float, float, float]:
+        if (token.name == self.base_token.name) & (type == "exact_output"):
+            # does the swap with the base token as output
+            if amount < self._base_token_amount:
+                _current_exchange_rate = self.exchange_rate()
+                _base_token_amount = self.base_token_amount - amount
+                _quote_token_amount = self.invariant / self._base_token_amount
+            else:
+                raise Exception(
+                    "Swap amount must be smaller than the amount of base tokens in the pool.")
+
+        elif (token.name == self.quote_token.name) & (type == "exact_output"):
```

Comment:
This function probably overcomplicates the logic flow rather than simplify it. `input`s have the same structure, and so do `output`s, so these parts could be standalone functions called directly by the other methods (such as `exact_input_swap_slippage`).

These two functions' signatures would look like: `def apply_input(input_token, passive_token) -> (new_input_token_amount, new_passive_token_amount)`.

Then you could have a `def compute_slippage(new_quote_amount, new_base_amount)` that computes the part at the end of this function:
```
        _new_exchange_rate = _quote_token_amount / _base_token_amount
        _slippage = float(_new_exchange_rate / self.exchange_rate() - 1)
```

Your usage of the returned parameters suggests these are separate bits that should not go together. Example from your code, where you wouldn't need to call `compute_slippage(...)` after refactoring as suggested:
```
        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
            token=input_token, amount=amount, type="exact_input"
        )
```

Reply:
I need some clarification to implement this. Lets discuss this sometimes.